UefiPayloadPkg/Library/PlatformBootManagerLib/PlatformBootManager.c

```diff
@@ -1065,6 +1065,142 @@ WarnIfBatteryLow (
   DrainInput ();
 }
 
+STATIC
+VOID
+WarnIfFirmwareUpdateMode (
+  VOID
+)
+{
+  EFI_STATUS     Status;
+  EFI_EVENT      TimerEvent;
+  EFI_EVENT      Events[2];
+  UINTN          Index;
+  EFI_INPUT_KEY  Key;
+  EFI_TIME       Time;
+  CHAR16         RandomDigit;
+  CHAR16         DelayLine[81];
+  CHAR16         PressKeyLine[81];
+  BOOLEAN        CursorVisible;
+  UINTN          CurrentAttribute;
+  UINTN          SecondsLeft;
+  UINTN          VarSize;
+  BOOLEAN        FUMEnabled;
+
+  VarSize = sizeof (FUMEnabled);
+
+  Status = gRT->GetVariable (
+      L"FirmwareUpdateMode",
+      &gDasharoSystemFeaturesGuid,
+      NULL,
+      &VarSize,
+      &FUMEnabled
+      );
+
+  if (EFI_ERROR(Status) || VarSize != sizeof(FUMEnabled) || !FUMEnabled) {
+    return;
+  }
+
+  //
+  // Remove variable to disable FUM on next boot
+  //
+  Status = gRT->SetVariable (
+      L"FirmwareUpdateMode",
+      &gDasharoSystemFeaturesGuid,
+      EFI_VARIABLE_BOOTSERVICE_ACCESS | EFI_VARIABLE_NON_VOLATILE,
+      0,
+      NULL
+      );
+
+  Status = gBS->CreateEvent (
+      EVT_TIMER,
+      TPL_CALLBACK,
+      NULL,
+      NULL,
+      &TimerEvent
+      );
+  ASSERT_EFI_ERROR (Status);
+
+  Status = gRT->GetTime (&Time, NULL);
+  //
+  // Don't check status, even if the call failed we still have "random" data
+  // from stack where Time is located. It is better than nothing, and we don't
+  // need more.
+  //
+  RandomDigit = L'0' + (Time.Second % 10);
+
+  UnicodeSPrint (
+      PressKeyLine,
+      sizeof (PressKeyLine),
+      L"Press %c to continue.",
+      RandomDigit
+      );
+
+  CurrentAttribute = gST->ConOut->Mode->Attribute;
+  CursorVisible    = gST->ConOut->Mode->CursorVisible;
+
+  gST->ConOut->EnableCursor (gST->ConOut, FALSE);
+
+  DrainInput ();
+  gBS->SetTimer (TimerEvent, TimerPeriodic, 1 * 1000 * 1000 * 10);
+
+  Events[0] = gST->ConIn->WaitForKey;
+  Events[1] = TimerEvent;
+
+  SecondsLeft = 30;
+  while (SecondsLeft > 0) {
+    UnicodeSPrint (
+        DelayLine,
+        sizeof (DelayLine),
+        L"(The platform will reboot automatically in %d second%a.)",

```

Comment:
```suggestion
        L"(The platform will reboot and disable Firmware Update Mode automatically in %d second%a.)",

```

Reply:
Done in https://github.com/Dasharo/edk2/compare/f83b4e816678d29b583601eaa698259b833fc3ec..270f5b14260c6bfc266400427b4cdad3c7593157. I split it to two lines, it was too long for one.